### PR TITLE
Added Cake build script.

### PR DESCRIPTION
**Pull request by patriksvensson**
Things left to do:

- [ ] Add Bash bootstrapper
- [x] Delete symbol packages
- [x] Generate NuGet packages
- [x] Publish to NuGet Gallery and Symbol Source
- [x] Publish to custom source
- [x] Tag current release
- [x] Update version (task)
- [x] Update informational version (task)
- [x] Update version and tag release.
- [x] Update to Cake 0.12.0 when released (for dotnet support)
- [x] Restore/Build/Test using `dotnet`.
- [x] Generate XML documentation in project.json.
- [x] Include `netstandard1.5` in zip file.
- [x] Include `netstandard1.5` in NuGet packages.

Not sure about naming of the `cake-build.ps1` file yet (since `build.ps1` already exist), but this decision can be postponed until later.

### Usage

Build:
```ps
.\build.ps1
```

Build without cleaning and skip running the tests (useful for debugging the build script).

```ps
.\build.ps1 -SkipClean -SkipTests
```

Update versions:
```ps
.\build.ps1 -Target Update-Version -Version "2.0.0"
.\build.ps1 -Target Update-Informational-Version -Version "2.0.0-beta"
```

Publish to nuget.org:
```ps
.\build.ps1 -Target Publish-NuGet -NuGetApiKey "KEY"
```

Publish to custom NuGet source:
```ps
.\build.ps1 -Target Publish-NuGet -NuGetApiKey "KEY" -NuGetSource "https://packages.com"
```

Prepare release
```ps
.\build.ps1 -Target Prepare-Release -Version "1.2.3"
```

Tag (with version i SharedAssemblyInfo):
```ps
.\build.ps1 -Target Tag
```

### CODE REVIEW

**Comment by jchannon** [comment]
Nice

**Comment by grumpydev** [comment]
Very nice.. Not looked too much at cake, but that's nice a d clean

**Comment by patriksvensson** [comment]
@jchannon @grumpydev Happy to hear that :smile:

**Comment by psibernetic** [comment]
Cake is great. Super happy about this.

**Comment by patriksvensson** [comment]
Ok, now the solution is restored, built and tested using the `dotnet` command. Only tagging left now.

Ping @thecodejunkie

**Comment by khellang** [comment]
Ideally, this PR should replace the current build file so we can actually see it build :smile:

File: build.cake [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    DotNetCoreBuild(project.GetDirectory().FullPath, new DotNetCoreBuildSettings {
+      Configuration = configuration,
+      Verbose = false
+    });
+  }
+});
+
+Task("Test")
+  .Description("Executes xUnit tests")
+  .WithCriteria(false)
```
Comment (khellang) [inline build.cake]: Is this supposed to be here?

File: build.cake [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    });
+  }
+});
+
+Task("Publish")
+  .Description("Gathers output files and copies them to the output folder")
+  .IsDependentOn("Compile")
+  .Does(() =>
+{
+  var files = GetFiles("src/**/bin/" + configuration + "/net452/*.dll")
```
Comment (khellang) [inline build.cake]: Ideally we'd want the `netstandard` (and any other target we have) to also be in here.

File: .gitignore [hunk 3]
```diff
@@ -1,34 +1,35 @@
...
-src/_NCrunch_Nancy/
-Thumbs.db
-.idea
-*.GhostDoc.xml
-Gemfile.lock
-.vs/
-packages/
-project.lock.json
-TestAssembly.dll
-.vscode/
```
Comment (khellang) [inline .gitignore]: What's up with the line endings?

File: build.cake [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+  nuspecs = GetFiles(outputNuGet.Path.FullPath + "/*.nuspec");
+  
+  // Update all nuspecs.
+  foreach(var nuspec in nuspecs) 
+  {
+    Information("Updating {0}...", nuspec.GetFilename().FullPath);
+    var document = XDocument.Load(nuspec.FullPath);
+    var ns = new XmlNamespaceManager(new NameTable());
+    ns.AddNamespace("ns", "http://schemas.microsoft.com/packaging/2010/07/nuspec.xsd");
+    document.XPathSelectElement("//ns:version", ns).Value = nancyVersion;
```
Comment (khellang) [inline build.cake]: I think we need to make this into a separate tool/method/something, cause it's an insane amount of noise. Does Cake already have something for patching nuspecs?

File: build.cake [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+using System.Xml;
+using System.Xml.Linq;
+using System.Xml.XPath;
+
+// Arguments
+var target = Argument<string>("target", "Default");
+var source = Argument<string>("source", null);
+var apiKey = Argument<string>("apikey", null);
+var version = Argument<string>("targetversion", null);
+var incremental = Argument<bool>("incremental", false);
```
Comment (khellang) [inline build.cake]: What does `incremental` mean? It seems to be switching on/off cleaning?

File: build.cake (hunk 1 above)
Comment (patriksvensson) [inline build.cake]: Nope

File: build.cake (hunk 2 above)
Comment (patriksvensson) [inline build.cake]: Yes, that's the next step.

File: .gitignore (hunk 3 above)
Comment (patriksvensson) [inline .gitignore]: Not sure to be honest.

File: build.cake (hunk 5 above)
Comment (patriksvensson) [inline build.cake]: Yes. Should probably be called "skipclean" or something like that. This is just something I use myself at the moment.

File: build.cake (hunk 4 above)
Comment (patriksvensson) [inline build.cake]: Yes. I've created an issue for this here: https://github.com/cake-build/cake/issues/916
Meanwhile, I will see if I can make this better.

**Comment by patriksvensson** [comment]
@khellang Replaced build.ps1 :smile:

**Comment by patriksvensson** [comment]
@jchannon @grumpydev @khellang @thecodejunkie This is ready for review.

Make sure to test this out on as many machines as you can and try to break it. I haven't encountered any bugs, but it might be possible that I've misunderstood something in the original script.

File: src/Nancy.Authentication.Basic/Nancy.Authentication.Basic.xproj [hunk 6]
```diff
@@ -10,6 +10,7 @@
     <RootNamespace>Nancy.Authentication.Basic</RootNamespace>
     <BaseIntermediateOutputPath Condition="'$(BaseIntermediateOutputPath)'=='' ">.\obj</BaseIntermediateOutputPath>
     <OutputPath Condition="'$(OutputPath)'=='' ">.\bin\</OutputPath>
+    <DocumentationFile>Nancy.xml</DocumentationFile>
```
Comment (patriksvensson) [inline src/Nancy.Authentication.Basic/Nancy.Authentication.Basic.xproj]: Whoopsie, this should be removed.

File: build.ps1 [hunk 7]
```diff
@@ -1,81 +1,101 @@
...
+    [switch]$SkipClean,
+    [switch]$SkipTests
+)
+
+######################################################################################################
+
+Function Install-Dotnet()
+{
+  $GlobalDotNetFound = &where.exe dotnet 2>&1 | Out-Null
+  $LocalDotNetFound = Test-Path (Join-Path $PSScriptRoot ".dotnet")
```
Comment (patriksvensson) [inline build.ps1]: Should make sure `$PSScriptRoot` has been set properly to work with older versions of PowerShell.

**Comment by devlead** [comment]
@jchannon build.sh and fixes now present, rebased with latest master. Please review.

**Comment by thecodejunkie** [comment]
A word of caution - there are a trillion commits going into this now with changes and stuff all over the place. Please document the changes and WHY they're needed or it will be a huge nightmare to both review the pull-request and support it by the team (because we won't know why half of the things were needed to be changed). I recommend you maintain a list of changes. PS. I am talking about the commits that @jchannon are pushing

**Comment by jchannon** [comment]
They were squashed

**Comment by thecodejunkie** [comment]
It's not the number of commits that worry me.. it's the amount of changes

**Comment by jchannon** [comment]
This was merged with Appveyor failing because it uses a more recent version of net core, TeamCity doesnt have net core and travis passed

**Pull request merged**